File: app/src/main/java/com/nirbhor/app/ui/screens/family/RefillScreen.kt
package com.nirbhor.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nirbhor.app.domain.Medicine
import com.nirbhor.app.domain.StockStatus
import com.nirbhor.app.domain.StockCalculator
import com.nirbhor.app.navigation.LocalAppContainer
import com.nirbhor.app.navigation.NavActions
import com.nirbhor.app.ui.components.NbCard
import com.nirbhor.app.ui.components.PrimaryButton
import com.nirbhor.app.ui.components.QuantityStepper
import com.nirbhor.app.ui.components.QuickChip
import com.nirbhor.app.ui.components.Scrim
import com.nirbhor.app.ui.components.SectionLabel
import com.nirbhor.app.ui.components.SheetSurface
import com.nirbhor.app.ui.components.StatusPill
import com.nirbhor.app.ui.components.TintPanel
import com.nirbhor.app.ui.components.NirbhorTopBar
import com.nirbhor.app.ui.i18n.num
import com.nirbhor.app.ui.i18n.tr
import com.nirbhor.app.ui.marks.MedicineMark
import com.nirbhor.app.ui.theme.Dimens
import com.nirbhor.app.ui.theme.NirbhorTheme
import kotlinx.coroutines.launch

/** Refill & stock (2u/2g) with the restock sheet (6e). */
@Composable
fun RefillScreen(actions: NavActions) {
    val colors = NirbhorTheme.colors
    val repo = LocalAppContainer.current.repository
    val scope = rememberCoroutineScope()
    val medicines by repo.medicines.collectAsStateWithLifecycle(initialValue = emptyList())
    val stock by repo.stockStatuses().collectAsStateWithLifecycle(initialValue = emptyList())
    val stockById = remember(stock) { stock.associateBy { it.medicineId } }

    var restockFor by remember { mutableStateOf<Medicine?>(null) }
    val lowest = medicines.minByOrNull { stockById[it.id]?.daysRemaining ?: Int.MAX_VALUE }
    val lowestStock = lowest?.let { stockById[it.id] }

    Box(Modifier.fillMaxSize()) {
        Column(Modifier.fillMaxSize().background(colors.paper)) {
            NirbhorTopBar(title = tr("রিফিল ও মজুত", "Refill & stock"), onBack = actions::back)
            Column(
                Modifier.fillMaxSize().verticalScroll(rememberScrollState())
                    .navigationBarsPadding().padding(horizontal = Dimens.screenPadding, vertical = 16.dp),
                verticalArrangement = Arrangement.spacedBy(Dimens.groupGap),
            ) {
                if (lowest != null && lowestStock != null && lowestStock.isLow) {
                    LowAlertCard(lowest, lowestStock) { restockFor = lowest }
                }
                Column(verticalArrangement = Arrangement.spacedBy(Dimens.cardGap)) {
                    if (medicines.isEmpty()) {
                        Text(tr("এখনও কোনো ওষুধ যোগ করা হয়নি।", "No medicines have been added yet."), style = NirbhorTheme.type.body, color = colors.ink2)
                        PrimaryButton(tr("ওষুধ যোগ করুন", "Add a medicine"), actions::startAddMedicine, modifier = Modifier.fillMaxWidth())
                    } else {
                        medicines.forEach { m -> StockRow(m, stockById[m.id]) { restockFor = m } }
                    }
                }
                TintPanel(background = colors.sage) {
                    Text(
                        tr("প্রতিটি ওষুধ কত দিন চলবে তা রোজ হিসাব করা হয়। ৫ দিনের কম থাকলে সতর্কবার্তা যায়।",
                            "Days remaining is recalculated daily. You're warned when fewer than 5 days are left."),
                        style = NirbhorTheme.type.body, color = colors.ink2,
                    )
                }
            }
        }

        val target = restockFor
        if (target != null) {
            Scrim(onDismiss = { restockFor = null })
            Box(Modifier.align(Alignment.BottomCenter)) {
                RestockSheet(
                    medicine = target,
                    current = stockById[target.id]?.count ?: 0,
                    onAdd = { amount ->
                        scope.launch { repo.addStock(target.id, amount) }
                        restockFor = null
                    },
                )
            }
        }
    }
}

@Composable
private fun LowAlertCard(med: Medicine, stock: StockStatus, onRestock: () -> Unit) {
    val colors = NirbhorTheme.colors
    Column(
        Modifier.fillMaxWidth().clip(RoundedCornerShape(Dimens.radiusLargeCard)).background(colors.warmSoft)
            .border(1.dp, colors.warm.copy(alpha = 0.28f), RoundedCornerShape(Dimens.radiusLargeCard)).padding(16.dp),
    ) {
        SectionLabel(tr("ফুরিয়ে আসছে", "Running low"), color = colors.warmD)
        Spacer(Modifier.height(10.dp))
        Row(verticalAlignment = Alignment.CenterVertically) {
            MedicineMark(shape = med.mark, color = Color(med.markColor), size = 34.dp)
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Text(med.displayName, style = NirbhorTheme.type.cardTitlePrimary, color = colors.ink)
                Text(
                    tr("${num(stock.count)}টি ${med.form} বাকি · ${num(stock.daysRemaining)} দিন চলবে",
                        "${stock.count} ${med.form} left · ${stock.daysRemaining} days"),
                    style = NirbhorTheme.type.meta, color = colors.warmD,
                )
            }
        }
        Spacer(Modifier.height(12.dp))
        SegmentBar(filled = stock.count.coerceIn(0, 10))
        Spacer(Modifier.height(12.dp))
        PrimaryButton(
            tr("আরও কিনেছি", "Bought more"), onRestock, height = 58.dp,
            container = colors.warmD, content = colors.paper, modifier = Modifier.fillMaxWidth(),
        )
    }
}

@Composable
private fun SegmentBar(filled: Int) {
    val colors = NirbhorTheme.colors
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(4.dp)) {
        repeat(10) { i ->
            Box(
                Modifier.weight(1f).height(12.dp).clip(RoundedCornerShape(6.dp))
                    .background(if (i < filled) colors.warm else colors.warm.copy(alpha = 0.2f)),
            )
        }
    }
}

@Composable
private fun StockRow(med: Medicine, stock: StockStatus?, onRestock: () -> Unit) {
    val colors = NirbhorTheme.colors
    val days = stock?.daysRemaining ?: 0
    val low = stock?.isLow == true
    NbCard {
        Row(verticalAlignment = Alignment.CenterVertically) {
            MedicineMark(shape = med.mark, color = Color(med.markColor), size = 32.dp)
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Text(med.displayName, style = NirbhorTheme.type.cardTitleSecondary, color = colors.ink)
                Text(
                    tr("${num(stock?.count ?: 0)}টি ${med.form} · ${num(days)} দিন চলবে",
                        "${stock?.count ?: 0} ${med.form} · ${days} days"),
                    style = NirbhorTheme.type.meta, color = colors.ink3,
                )
            }
            if (low) StatusPill(tr("শীঘ্রই", "Soon"), colors.warmSoft, colors.warmD)
            else StatusPill(tr("ঠিক আছে", "OK"), colors.calmSoft, colors.calmD)
        }
        Spacer(Modifier.height(10.dp))
        Box(Modifier.fillMaxWidth().height(8.dp).clip(RoundedCornerShape(4.dp)).background(colors.sage)) {
            val frac = (days / 30f).coerceIn(0f, 1f)
            Box(Modifier.fillMaxWidth(frac).height(8.dp).clip(RoundedCornerShape(4.dp)).background(if (low) colors.warm else colors.calm))
        }
        Spacer(Modifier.height(10.dp))
        PrimaryButton(tr("মজুত আপডেট করুন", "Update stock"), onRestock, height = 48.dp, modifier = Modifier.fillMaxWidth())
    }
}

@Composable
private fun RestockSheet(medicine: Medicine, current: Int, onAdd: (Int) -> Unit) {
    val colors = NirbhorTheme.colors
    var amount by remember { mutableFloatStateOf(30f) }
    SheetSurface {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            MedicineMark(shape = medicine.mark, color = Color(medicine.markColor), size = 34.dp)
            Spacer(Modifier.width(12.dp))
            Text(medicine.displayName, style = NirbhorTheme.type.cardTitlePrimary, color = colors.ink)
        }
        Spacer(Modifier.height(16.dp))
        QuantityStepper(
            value = amount, onChange = { amount = it },
            valueLabel = num(amount.toInt()), unitLabel = tr("টি যোগ করুন", "to add"),
            size = 72.dp, step = 1f,
        )
        Spacer(Modifier.height(12.dp))
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            listOf(10, 30, 60, 90).forEach { q ->
                QuickChip(num(q), selected = amount.toInt() == q, onClick = { amount = q.toFloat() }, modifier = Modifier.weight(1f))
            }
        }
        Spacer(Modifier.height(14.dp))
        val result = current + amount.toInt()
        val days = StockCalculator.estimatedDays(
            result, medicine.dosePerIntake, medicine.timeTokens.size,
            medicine.frequency, medicine.weekdaysMask, medicine.paused,
        )
        TintPanel(background = colors.calmSoft) {
            Text(
                if (days == null) tr("ঘরে হবে ${num(result)}টি — ওষুধটি এখন বন্ধ আছে।", "You'll have $result — this medicine is paused.")
                else tr("ঘরে হবে ${num(result)}টি — প্রায় ${num(days)} দিন চলবে।", "You'll have ${result} — about ${days} days."),
                style = NirbhorTheme.type.body, color = colors.calmD,
            )
        }
        Spacer(Modifier.height(14.dp))
        PrimaryButton(tr("যোগ করুন", "Add"), { onAdd(amount.toInt()) }, height = 64.dp, modifier = Modifier.fillMaxWidth())
    }
}

File: lib/ui/screens/refill_screen.dart
import 'package:flutter/material.dart';

import '../../data/app_scope.dart';
import '../../domain/models.dart';
import '../../domain/stock_calculator.dart';
import '../../navigation/nav_actions.dart';
import '../../theme/theme.dart';
import '../components/buttons.dart';
import '../components/controls.dart';
import '../components/labels.dart';
import '../components/overlays.dart';
import '../components/scaffold.dart';
import '../components/surfaces.dart';
import '../marks/medicine_mark.dart';

/// Refill & stock (2u/2g) — the low-stock alert, the per-medicine bars, and the restock sheet.
class RefillScreen extends StatelessWidget {
  const RefillScreen({super.key});

  Future<void> _restock(BuildContext context, Medicine med, int current) async {
    final added = await showNbSheet<int>(
      context,
      (context) => _RestockSheet(medicine: med, current: current),
    );
    if (added != null && added > 0 && context.mounted) {
      await context.repo.addStock(med.id, added);
    }
  }

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return Scaffold(
      backgroundColor: colors.paper,
      body: Column(
        children: [
          NirbhorTopBar(
            title: context.tr('রিফিল ও মজুত', 'Refill & stock'),
            onBack: context.nav.back,
          ),
          Expanded(
            child: RepoBuilder<(List<Medicine>, List<StockStatus>)>(
              query: (repo) async => (await repo.medicines(), await repo.stockStatuses()),
              builder: (context, data) {
                final (medicines, stock) = data;
                final stockById = {for (final s in stock) s.medicineId: s};

                Medicine? lowest;
                var lowestDays = 1 << 30;
                for (final m in medicines) {
                  final d = stockById[m.id]?.daysRemaining ?? (1 << 30);
                  if (d < lowestDays) {
                    lowestDays = d;
                    lowest = m;
                  }
                }
                final lowestStock = lowest == null ? null : stockById[lowest.id];

                return SingleChildScrollView(
                  padding: const EdgeInsets.symmetric(
                    horizontal: Dimens.screenPadding,
                    vertical: 16,
                  ),
                  child: SafeArea(
                    top: false,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.stretch,
                      children: [
                        if (lowest != null && lowestStock != null && lowestStock.isLow) ...[
                          _LowAlertCard(
                            med: lowest,
                            stock: lowestStock,
                            onRestock: () => _restock(context, lowest!, lowestStock.count),
                          ),
                          const SizedBox(height: Dimens.groupGap),
                        ],
                        if (medicines.isEmpty) ...[
                          Text(
                            context.tr(
                              'এখনও কোনো ওষুধ যোগ করা হয়নি।',
                              'No medicines have been added yet.',
                            ),
                            style: context.type.body.copyWith(color: colors.ink2),
                          ),
                          const SizedBox(height: Dimens.cardGap),
                          PrimaryButton(
                            text: context.tr('ওষুধ যোগ করুন', 'Add a medicine'),
                            onPressed: context.nav.startAddMedicine,
                          ),
                        ] else
                          for (final m in medicines) ...[
                            _StockRow(
                              med: m,
                              stock: stockById[m.id],
                              onRestock: () =>
                                  _restock(context, m, stockById[m.id]?.count ?? 0),
                            ),
                            const SizedBox(height: Dimens.cardGap),
                          ],
                        const SizedBox(height: Dimens.groupGap),
                        TintPanel(
                          background: colors.sage,
                          child: Text(
                            context.tr(
                              'প্রতিটি ওষুধ কত দিন চলবে তা রোজ হিসাব করা হয়। ৫ দিনের কম থাকলে সতর্কবার্তা যায়।',
                              "Days remaining is recalculated daily. You're warned when fewer than 5 days are left.",
                            ),
                            style: context.type.body.copyWith(color: colors.ink2),
                          ),
                        ),
                      ],
                    ),
                  ),
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

class _LowAlertCard extends StatelessWidget {
  const _LowAlertCard({required this.med, required this.stock, required this.onRestock});

  final Medicine med;
  final StockStatus stock;
  final VoidCallback onRestock;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final remainingDays = stock.daysRemaining ?? 0;
    return Container(
      decoration: BoxDecoration(
        color: colors.warmSoft,
        borderRadius: BorderRadius.circular(Dimens.radiusLargeCard),
        border: Border.all(color: colors.warm.withValues(alpha: 0.28)),
      ),
      padding: const EdgeInsets.all(16),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          SectionLabel(context.tr('ফুরিয়ে আসছে', 'Running low'), color: colors.warmD),
          const SizedBox(height: 10),
          Row(
            children: [
              MedicineMark(shape: med.mark, color: Color(med.markColor), size: 34),
              const SizedBox(width: 12),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      med.displayName,
                      style: context.type.cardTitlePrimary.copyWith(color: colors.ink),
                    ),
                    Text(
                      context.tr(
                        '${context.num(stock.count)}টি ${med.form} বাকি · ${context.num(remainingDays)} দিন চলবে',
                        '${stock.count} ${med.form} left · $remainingDays days',
                      ),
                      style: context.type.meta.copyWith(color: colors.warmD),
                    ),
                  ],
                ),
              ),
            ],
          ),
          const SizedBox(height: 12),
          _SegmentBar(filled: stock.count.clamp(0, 10)),
          const SizedBox(height: 12),
          PrimaryButton(
            text: context.tr('আরও কিনেছি', 'Bought more'),
            onPressed: onRestock,
            height: 58,
            container: colors.warmD,
            content: colors.paper,
          ),
        ],
      ),
    );
  }
}

class _SegmentBar extends StatelessWidget {
  const _SegmentBar({required this.filled});

  final int filled;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return Row(
      children: [
        for (var i = 0; i < 10; i++) ...[
          if (i > 0) const SizedBox(width: 4),
          Expanded(
            child: Container(
              height: 12,
              decoration: BoxDecoration(
                color: i < filled ? colors.warm : colors.warm.withValues(alpha: 0.2),
                borderRadius: BorderRadius.circular(6),
              ),
            ),
          ),
        ],
      ],
    );
  }
}

class _StockRow extends StatelessWidget {
  const _StockRow({required this.med, required this.stock, required this.onRestock});

  final Medicine med;
  final StockStatus? stock;
  final VoidCallback onRestock;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    // null means "no rate to divide by" (paused, or no times set) — never an infinite supply.
    final days = stock?.daysRemaining;
    final low = stock?.isLow ?? false;
    final count = stock?.count ?? 0;

    return NbCard(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Row(
            children: [
              MedicineMark(shape: med.mark, color: Color(med.markColor), size: 32),
              const SizedBox(width: 12),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      med.displayName,
                      style: context.type.cardTitleSecondary.copyWith(color: colors.ink),
                    ),
                    Text(
                      days == null
                          ? context.tr(
                              '${context.num(count)}টি ${med.form} · ${med.paused ? "এখন বন্ধ" : "সময় ঠিক করা নেই"}',
                              '$count ${med.form} · ${med.paused ? "paused" : "no times set"}',
                            )
                          : context.tr(
                              '${context.num(count)}টি ${med.form} · ${context.num(days)} দিন চলবে',
                              '$count ${med.form} · $days days',
                            ),
                      style: context.type.meta.copyWith(color: colors.ink3),
                    ),
                  ],
                ),
              ),
              if (low)
                StatusPill(
                  text: context.tr('শীঘ্রই', 'Soon'),
                  background: colors.warmSoft,
                  contentColor: colors.warmD,
                )
              else if (days == null)
                StatusPill(
                  text: context.tr('বন্ধ', 'Paused'),
                  background: colors.sage,
                  contentColor: colors.ink2,
                )
              else
                StatusPill(
                  text: context.tr('ঠিক আছে', 'OK'),
                  background: colors.calmSoft,
                  contentColor: colors.calmD,
                ),
            ],
          ),
          const SizedBox(height: 10),
          ClipRRect(
            borderRadius: BorderRadius.circular(4),
            child: SizedBox(
              height: 8,
              child: ColoredBox(
                color: colors.sage,
                child: FractionallySizedBox(
                  alignment: Alignment.centerLeft,
                  widthFactor: ((days ?? 0) / 30).clamp(0.0, 1.0),
                  child: ColoredBox(color: low ? colors.warm : colors.calm),
                ),
              ),
            ),
          ),
          const SizedBox(height: 10),
          // The sheet only adds to the count; correcting it downward lives on the medicine screen.
          PrimaryButton(
            text: context.tr('আরও যোগ করুন', 'Add more'),
            onPressed: onRestock,
            height: 48,
          ),
        ],
      ),
    );
  }
}

class _RestockSheet extends StatefulWidget {
  const _RestockSheet({required this.medicine, required this.current});

  final Medicine medicine;
  final int current;

  @override
  State<_RestockSheet> createState() => _RestockSheetState();
}

class _RestockSheetState extends State<_RestockSheet> {
  double _amount = 30;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final medicine = widget.medicine;
    final result = widget.current + _amount.toInt();
    final days = StockCalculator.estimatedDaysFrom(
      stockCount: result,
      dosePerIntake: medicine.dosePerIntake,
      timesPerScheduledDay: medicine.timeTokens.length,
      frequency: medicine.frequency,
      weekdaysMask: medicine.weekdaysMask,
      paused: medicine.paused,
    );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      mainAxisSize: MainAxisSize.min,
      children: [
        Row(
          children: [
            MedicineMark(shape: medicine.mark, color: Color(medicine.markColor), size: 34),
            const SizedBox(width: 12),
            Expanded(
              child: Text(
                medicine.displayName,
                style: context.type.cardTitlePrimary.copyWith(color: colors.ink),
              ),
            ),
          ],
        ),
        const SizedBox(height: 16),
        QuantityStepper(
          value: _amount,
          onChanged: (v) => setState(() => _amount = v.clamp(1, 10000)),
          valueLabel: context.num(_amount.toInt()),
          unitLabel: context.tr('টি যোগ করুন', 'to add'),
          step: 1,
          min: 1,
        ),
        const SizedBox(height: 12),
        Row(
          children: [
            for (final q in const [10, 30, 60, 90]) ...[
              if (q != 10) const SizedBox(width: 8),
              Expanded(
                child: QuickChip(
                  label: context.num(q),
                  selected: _amount.toInt() == q,
                  onTap: () => setState(() => _amount = q.toDouble()),
                ),
              ),
            ],
          ],
        ),
        const SizedBox(height: 14),
        TintPanel(
          background: colors.calmSoft,
          child: Text(
            days == null
                ? context.tr(
                    'ঘরে হবে ${context.num(result)}টি — ওষুধটি এখন বন্ধ আছে।',
                    "You'll have $result — this medicine is paused.",
                  )
                : context.tr(
                    'ঘরে হবে ${context.num(result)}টি — প্রায় ${context.num(days)} দিন চলবে।',
                    "You'll have $result — about $days days.",
                  ),
            style: context.type.body.copyWith(color: colors.calmD),
          ),
        ),
        const SizedBox(height: 14),
        PrimaryButton(
          text: context.tr('যোগ করুন', 'Add'),
          onPressed: () => Navigator.of(context).pop(_amount.toInt()),
        ),
      ],
    );
  }
}
